### PR TITLE
util: Test

### PR DESCRIPTION
**Pull request by ahmad-abuziad**
The goal of this PR to test cover util/util.go and increase it from 64.8% to 100% since almost all the util functions are testable
Screenshot From 2025-01-22 13-43-38

### CODE REVIEW

**Comment by purpleidea** [comment]
These will keep failing because of the commit message. Look at the tests. Lmk if you need more help.

**Comment by ahmad-abuziad** [comment]
> These will keep failing because of the commit message. Look at the tests. Lmk if you need more help.

I will look into it.
I will stick with the format since i started with it, except the last commit will contain a message that doesn't fail.

**Comment by purpleidea** [comment]
You can easily rename a bunch with: EDITOR=vim git rebase -i <commit id before the first one>

**Comment by ahmad-abuziad** [comment]
coverage increased from 64.8% to 94.4
image

File: util/util_test.go [hunk 1]
```diff
@@ -1636,3 +1639,960 @@ func TestSortMapStringValuesByUInt64Keys(t *testing.T) {
...
+			input: []string{},
+			want:  "",
+		},
+		{
+			name:  "single path",
+			input: []string{"/path/to"},
+			want:  "/path/to",
+		},
+		{
+			name:  "single path doesn't start with /", // TODO(ahmadabuziad): need to double check the desired output with @purple idea
```
Comment (purpleidea) [inline util/util_test.go:2139]: What should we do here? But if it's very important put `XXX: message` and remove your username and mine please

File: util/util_test.go [hunk 2]
```diff
@@ -1636,3 +1639,960 @@ func TestSortMapStringValuesByUInt64Keys(t *testing.T) {
...
+		if gotStr != tt.want.rebasedPath {
+			t.Errorf("got rebased path: %v, want rebased path: %v", gotStr, tt.want.rebasedPath)
+		}
+		if gotErr != tt.want.err {
+			t.Errorf("got error: %v, want error to be: %v", gotErr, tt.want.err)
+		}
+	}
+
+	t.Run("root doesn't end with /", func(t *testing.T) {
+		// TODO(ahmadabuziad) in Rebase return predefined error instead e.g.
```
Comment (purpleidea) [inline util/util_test.go:2253]: Remove username please.

File: util/util_test.go [hunk 3]
```diff
@@ -1636,3 +1639,960 @@ func TestSortMapStringValuesByUInt64Keys(t *testing.T) {
...
+		if gotStr != "" {
+			t.Errorf("got: %v, want empty string", gotStr)
+		}
+
+		if gotErr.Error() != "must be absolute" {
+			t.Errorf(`got error: %v, want error "must be absolute"`, gotErr.Error())
+		}
+	})
+
+	//TODO(ahmadabuziad): edge cases currently panic, handle edge cases. "/", ""
```
Comment (purpleidea) [inline util/util_test.go:2293]: Same as above...

**Review by purpleidea — CHANGES_REQUESTED** [review]
Thanks for all the tests.

For a first start, can you please rename the commit messages so they have the right format, typically `util: Test blah blah...` in this case.

Next see the other misc things, then we'll do a final review. Thank you!

Sorry for the delay, I've been at the FOSDEM+CfgMgmtCamp conference stuff.

File: util/util_test.go (hunk 1 above)
Comment (ahmad-abuziad) [reply util/util_test.go:2139]: - We should return empty string "" (undefined behavior) same as the next test case.
When sending multiple paths and one of them is not absolute the current behavior returns an empty string, so for the case when single path is not absolute it should return an empty string. (the change may break existing usage - it showed me there is no usage for this function)

- What does `XXX: message` format mean?
- removed usernames

**Comment by purpleidea** [comment]
I expect to still see failures, test is here: https://github.com/purpleidea/mgmt/blob/master/test/test-commit-message.sh#L17

So for example a valid commit is:

foo: Hello something

an invalid commit is:

foo: hello something

Note the capitalization.

**Comment by purpleidea** [comment]
Sorry for the delays... So this was very hard to review because I couldn't cherry-pick out anything without massive git conflicts because it was all in the same file...

So I had to squash everything down, leaving out a few problematic tests and with one small fixup.

Merged, thank you!

Note: I had to drop the time-based tests since they aren't guaranteed to not be flaky in slow CI. If I mis-read this please let me know

**Comment by purpleidea** [comment]
Btw nice catch on some of the corner cases. The interesting one is actually mentioned in the docs so I commented that out.
